bee-network/src/peers/list.rs

```diff
@@ -0,0 +1,217 @@
+            .0
+            .write()
+            .await
+            .remove(id)
+            .ok_or_else(|| Error::UnlistedPeer(id.short()))?;
+
+        Ok(info)
+    }
+
+    // TODO: batch messages before sending using 'send_all' (e.g. batch messages for like 50ms)
+    pub async fn send_message(&self, message: Vec<u8>, to: &PeerId) -> Result<(), Error> {
+        let mut this = self.0.write().await;
+        let (_, state) = this.get_mut(to).ok_or_else(|| Error::UnlistedPeer(to.short()))?;
+
+        if let PeerState::Connected(sender) = state {
+            sender
+                .send(message)
+                .map_err(|_| Error::SendMessageFailure(to.short()))?;
+
+            Ok(())
+        } else {
+            Err(Error::DisconnectedPeer(to.short()))
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn count(&self) -> usize {
+        self.0.read().await.len()
+    }
+
+    pub async fn get_info(&self, id: &PeerId) -> Result<PeerInfo, Error> {
+        self.0
+            .read()
+            .await
+            .get(id)
+            .ok_or_else(|| Error::UnlistedPeer(id.short()))
+            // .map(|kv| kv.value().0.clone())
+            .map(|kv| kv.0.clone())
+    }
+
+    pub async fn is(&self, id: &PeerId, predicate: impl Fn(&PeerInfo, &PeerState) -> bool) -> Result<bool, Error> {
+        self.0
+            .read()
+            .await
+            .get(id)
+            .ok_or_else(|| Error::UnlistedPeer(id.short()))
+            // .map(|kv| predicate(&kv.value().0, &kv.value().1))
+            .map(|kv| predicate(&kv.0, &kv.1))
+    }
+
+    pub async fn iter_if(&self, predicate: impl Fn(&PeerInfo, &PeerState) -> bool) -> impl Iterator<Item = PeerId> {
+        self.0
+            .read()
+            .await
+            .iter()
+            .filter_map(|kv| {
+                let (info, state) = kv.1; // kv.value();
+                if predicate(info, state) {
+                    // Some(kv.key().clone())
+                    Some(kv.0.clone())
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<PeerId>>()
+            .into_iter()
+    }
+
+    pub async fn count_if(&self, predicate: impl Fn(&PeerInfo, &PeerState) -> bool) -> usize {
+        self.0.read().await.iter().fold(0, |count, kv| {
+            // let (info, state) = kv.value();
+            let (info, state) = kv.1;
+            if predicate(info, state) {
+                count + 1
+            } else {
+                count
+            }
+        })
+    }
+
+    pub async fn remove_if(&self, id: &PeerId, predicate: impl Fn(&PeerInfo, &PeerState) -> bool) {
+        let remove = if let Some((info, state)) = self.0.read().await.get(id) {
+            predicate(info, state)
+        } else {
+            return;
+        };
+
+        if remove {
+            let _ = self.0.write().await.remove(id);
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn clear(&self) {
```

Comment:
In what circumstance would this occur? If it only ever occurs with synchronisation, you could change the method receiver to `&mut self` and use `self.0.get_mut().clear();`